kong/tools/request_aware_table.lua

```diff
@@ -0,0 +1,89 @@
+local table_clear = require "table.clear"
+
+local get_phase = ngx.get_phase
+local var       = ngx.var
+local fmt       = string.format
+
+local NGX_VAR_PHASES = {
+  set           = true,
+  rewrite       = true,
+  balancer      = true,
+  access        = true,
+  content       = true,
+  header_filter = true,
+  body_filter   = true,
+  log           = true,
+}
+
+
+--- Request aware table constructor
+-- Wraps an existing table (or creates a new one) with request-aware access
+-- logic to protect the underlying data from race conditions.
+-- @param data_table (optional) The target table to use as underlying data
+-- @return The newly created table with request-aware access
+local function new(data_table)
+  if data_table and type(data_table) ~= "table" then
+    error("data_table must be a table", 2)
+  end
+
+  local allowed_request_id
+  local proxy = {}
+  local data = data_table or {}
+  -- no need to consider the dynamic log level since the request-aware feature
+  -- is only for debugging purposes (not production use)
+  local is_debug_mode = kong.configuration.log_level == "debug"
```

Comment:
This should be cached at the module level, and the `is_debug_mode` check should be used very early in the `new` function to return early. There is no reason to even  create or return a patched table if `is_debug_mode` is `false` otherwise we introduce penalty even when debug mode is disabled.

Reply:
done. Thanks!